### ISSUE TITLE
PollEvents throws an Execution Engine Exception

### ISSUE THREAD

**Issue by KreideGit** [ForeverZer0/glfw-net#19]
I am currently trying to setup some callbacks for the mouse and keyboard. 
I have a very basic loop going: Draw, Swap buffers, Poll events. 
Everytime I move my mouse or spin the mouse wheel it prints the information onto the console. But after a certain amount of time the PollEvents function throws this exception and the only way to avoid this is to not use these events. Is this a bug or am I missing something?

**Comment by ForeverZer0** [comment]
Would you be able to share some of the relevant code, namely where the callbacks are set, the main loop where `PollEvents` is called? 

Please also indlude which framework you are using. There was a previous issue with binding events using lambda functions that caused errors on .NET Framework, but not on .NET Core, so it may possibly be relevant. From the sounds of it, there is a reference to either the window or a callback that is getting lost, and being eaten by the garbage collector, though it is hard to say anything definitive without some more info.

**Comment by KreideGit** [comment]
I'm currently using .NET Core 2.2.
This code runs in my Program.cs in the main function:
https://pastebin.com/3nUBU01f
The events just print out information like positions to the console.
This class is my own window class:
https://pastebin.com/GXvd5LDX
The import functions are PollEvents(), SetupMouseEvents() and SetupKeyboardEvents(). You can also see that I have wrote some EventHandler in order to access the callbacks more easily outside the class.  Am I doing something wrong when setting up these events?

**Comment by KreideGit** [comment]
SetupMouseEvents() Line: 205
SetupKeyboardEvents() Line: 238
PollEvents() Line: 88

**Comment by ForeverZer0** [comment]
> Am I doing something wrong when setting up these events?

Sort of, but only because of the nature of the callbacks being unmanaged code, not for traditional C# that is fully managed and does not call into a native library. 

To hopefully clarify, take this example:
```csharp
Glfw.SetScrollCallback(_window, (w, x, y) =>
{
    MouseWheelEventArgs e = new MouseWheelEventArgs(x, y);
    OnMouseWheelChanged?.Invoke(this, e);
});
```

When setting a callback, you are merely passing a pointer to a function that the unmanaged code will invoke when needed. The **unmanaged** code is storing this reference, but in your example, the **managed** code has no indication that the lamda function is needed any longer. It is not being stored in a variable (such as a delegate or event), so it gets garbage collected. It is not aware that lamda is still needed, since it is not managed (hence the name). Then the unmanaged code tries to call this function, but it is no longer valid and segfaults, causing your application to crash.

This behavior is unavoidable and true for all libraries that allow for setting callback functions, but there are multiple ways to still accomplish this. The static `Glfw` class is basically the "raw" bindings, without any of the sugar or safety nets to prevent this, hence the inclusion of the `NativeWindow` class to accommodate.

# TL;DR

## Solution 1
Create private delegate fields that store the references in managed code. This will prevent the garbage collector from stealing them. This is easiest illustrated in the `NativeWindow` class, specifically take a look at the `BindCallbacks` method. It stores a lamda in a delegate field, and then passes it to the callback function.

## Solution 2
Set the callback to a method, not a lamda. 
```csharp
public event EventHandler MouseWheelChange; // Disregard incorrect handler type...

protected virtual void OnMouseWheelChange(Window window, double x, double y) 
{
    MouseWheelChange?.Invoke(x, y);
}
```
Since the function pointer for a concrete-defined method will not get garbage collected, it is safe to set. The pointer to that function exists for the lifetime the object that contains it.

**Comment by KreideGit** [comment]
I have tried to set the callback to a protected virtual void function with the parameters of the event. I did this for every callback but it still throws the same exception. I also have to to store the callback with the lambda in a private variable declared in my class but this does not work either.

**Comment by ForeverZer0** [comment]
Have you tried using the `NativeWindow` class? From your examples, it appears to be pretty much what you are trying to implement, but handles these nuances for you. It is intended to be used as a base class to be inherited, think of it as what the `Form` class in the `System.Windows.Forms` namespace.  You can still also directly use the low-level GLFW class directly if desired, as it is basically a wrapper around a `Window` struct, and contains a field you can use for passing to the lower-level functions.

**Comment by ForeverZer0** [comment]
Here is minimal working example of Solution 2 from above, using the low-level functions.

```csharp
using System;
using GLFW;

namespace TestGLFW
{
    class Program
    {
        
        static void Main(string[] args)
        {
            var window = Glfw.CreateWindow(800, 600, "MyGame", Monitor.None, Window.None);
            Glfw.MakeContextCurrent(window);
            SetCallbacks(window);
            while (!Glfw.WindowShouldClose(window))
            {
                Glfw.SwapBuffers(window);
                Glfw.PollEvents();
            }
        }

        private static void SetCallbacks(Window window)
        {
            Glfw.SetCursorPositionCallback(window, MouseCallback);
        }

        private static void MouseCallback(IntPtr window, double x, double y)
        {
            Console.WriteLine($"<X: {x}, Y: {y}>");
        }
    }
}
```

**Comment by KreideGit** [comment]
I would like to learn GLFW myself that is why I do not use premade classes or abstraction even though they are way better than mine probably. Thankfully it is working now because I took a look at your NativeWindow class + I know what I did wrong. I was constructing the callbacks inside the function SetCallback, so SetCallback(w, new Callback(Function));. I think that the new Callback got garbage collected and that is the reason why it was throwing the exception over and over. Thanks again for the clear explanation.

**Comment by ForeverZer0** [comment]
Totally understand preferring not use abstractions, I am of a similar mindset. 

Essentially you just have to be careful when sending a pointer/reference/delegate into unmanaged code, and need to keep the same reference stored in managed code to prevent garbage collection. 

Glad you got working.
